File: dds/src/dcps/publication/data_writer_listener.rs
//! DataWriterListener - Callback interface for DataWriter status events.
//!
//! The `DataWriterListener` trait allows applications to receive asynchronous notifications
//! about status changes related to a `DataWriter`. Listeners are registered when creating
//! a DataWriter or by calling `DataWriter::set_listener()`.
//!
//! # Available Callbacks
//!
//! - `on_offered_deadline_missed`: Called when the writer fails to write data within the deadline period
//! - `on_offered_incompatible_qos`: Called when a matching reader has incompatible QoS
//! - `on_liveliness_lost`: Called when the writer fails to assert its liveliness
//! - `on_publication_matched`: Called when a matching subscription is discovered or lost

use crate::infrastructure::status::{
    LivelinessLostStatus, OfferedDeadlineMissedStatus, OfferedIncompatibleQosStatus,
    PublicationMatchedStatus,
};

use super::data_writer::DataWriter;

pub trait DataWriterListener: 'static + Send + Sync {
    type Foo;
    fn on_offered_deadline_missed(
        &self,
        _writer: &DataWriter<Self::Foo>,
        _status: &OfferedDeadlineMissedStatus,
    ) {
    }
    fn on_offered_incompatible_qos(
        &self,
        _writer: &DataWriter<Self::Foo>,
        _status: &OfferedIncompatibleQosStatus,
    ) {
    }
    fn on_liveliness_lost(&self, _writer: &DataWriter<Self::Foo>, _status: &LivelinessLostStatus) {}
    fn on_publication_matched(
        &self,
        _writer: &DataWriter<Self::Foo>,
        _status: &PublicationMatchedStatus,
    ) {
    }
}

File: java/api/src/main/java/com/intellectus/int2dds/listeners/DataWriterListener.java
package com.intellectus.int2dds.listeners;

import com.intellectus.int2dds.core.DataWriter;
import com.intellectus.int2dds.status.LivelinessLostStatus;
import com.intellectus.int2dds.status.OfferedDeadlineMissedStatus;
import com.intellectus.int2dds.status.OfferedIncompatibleQosStatus;
import com.intellectus.int2dds.status.PublicationMatchedStatus;

/**
 * Receives status-change notifications for a {@link DataWriter}.
 *
 * <p>Callbacks fire on native DDS background threads, not the caller's thread —
 * implementations must be thread-safe. Any exception thrown from a callback is
 * described and cleared at the native boundary and never propagates back into
 * the DDS thread.
 *
 * <p><b>v1 limitation:</b> the {@code writer} argument is always {@code null}.
 * Mapping a native handle back to its Java {@link DataWriter} entity is out of
 * scope for this branch; most listeners need only the status. Prefer extending
 * {@link DataWriterListenerBase} so future callbacks stay source-compatible.
 */
public interface DataWriterListener {

    /**
     * Invoked when the set of matched DataReaders changes.
     *
     * @param writer always {@code null} in v1 (see the interface note)
     * @param status the matched-reader counts at the time of the change
     */
    void onPublicationMatched(DataWriter<?> writer, PublicationMatchedStatus status);

    /**
     * Invoked when the writer misses the deadline it offered for an instance.
     *
     * @param writer always {@code null} in v1 (see the interface note)
     * @param status the missed-deadline counts at the time of the event
     */
    void onOfferedDeadlineMissed(DataWriter<?> writer, OfferedDeadlineMissedStatus status);

    /**
     * Invoked when the writer matches a DataReader with an incompatible QoS.
     *
     * @param writer always {@code null} in v1 (see the interface note)
     * @param status the incompatible-QoS counts at the time of the event
     */
    void onOfferedIncompatibleQos(DataWriter<?> writer, OfferedIncompatibleQosStatus status);

    /**
     * Invoked when the writer fails to signal its liveliness within its offered period.
     *
     * @param writer always {@code null} in v1 (see the interface note)
     * @param status the liveliness-lost counts at the time of the event
     */
    void onLivelinessLost(DataWriter<?> writer, LivelinessLostStatus status);
}
